executor/cte.go

```diff
@@ -0,0 +1,434 @@
+		if err = Next(ctx, e.recursiveExec, chk); err != nil {
+			return err
+		}
+		if chk.NumRows() == 0 {
+			if err = e.setupTblsForNewIteration(); err != nil {
+				return err
+			}
+			if e.iterInTbl.NumChunks() == 0 {
+				break
+			}
+			// Next iteration begins. Need use iterOutTbl as input of next iteration.
+			e.curIter++
+			e.iterInTbl.SetIter(e.curIter)
+			if e.curIter > e.ctx.GetSessionVars().CTEMaxRecursionDepth {
+				return ErrCTEMaxRecursionDepth.GenWithStackByArgs(e.curIter)
+			}
+			// Make sure iterInTbl is setup before Close/Open,
+			// because some executors will read iterInTbl in Open() (like IndexLookupJoin).
+			if err = e.recursiveExec.Close(); err != nil {
+				return err
+			}
+			if err = e.recursiveExec.Open(ctx); err != nil {
+				return err
+			}
+		} else {
+			if err = e.iterOutTbl.Add(chk); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (e *CTEExec) setupTblsForNewIteration() (err error) {
+	num := e.iterOutTbl.NumChunks()
+	chks := make([]*chunk.Chunk, 0, num)
+	// Setup resTbl's data.
+	for i := 0; i < num; i++ {
+		chk, err := e.iterOutTbl.GetChunk(i)
+		if err != nil {
+			return err
+		}
+		chk, err = e.tryFilterAndAdd(chk, e.resTbl, e.resHashTbl)
+		if err != nil {
+			return err
+		}
+		chks = append(chks, chk)
+	}
+
+	// Setup new iteration data in iterInTbl.
+	if err = e.iterInTbl.Reopen(); err != nil {
+		return err
+	}
+	if e.isDistinct {
+		for _, chk := range chks {
+			if _, err = e.tryFilterAndAdd(chk, e.iterInTbl, e.iterInHashTbl); err != nil {
+				return err
+			}
+		}
+	} else {
+		if err = e.iterInTbl.SwapData(e.iterOutTbl); err != nil {
+			return err
+		}
+	}
+	close(e.iterInTbl.GetBegCh())
+
+	// Clear data in iterOutTbl.
+	return e.iterOutTbl.Reopen()
+}
+
+func (e *CTEExec) reset() {
+	e.curIter = 0
+	e.chkIdx = 0
+	e.iterInHashTbl = nil
+	e.resHashTbl = nil
+}
+
+func (e *CTEExec) reopenTbls() (err error) {
+	if err := e.resTbl.Reopen(); err != nil {
+		return err
+	}
+	return e.iterInTbl.Reopen()
+}
+
+func setupCTEStorageTracker(tbl cteutil.Storage, ctx sessionctx.Context) {
+	memTracker := tbl.GetMemTracker()
+	memTracker.SetLabel(memory.LabelForCTEStorage)
+	memTracker.AttachTo(ctx.GetSessionVars().StmtCtx.MemTracker)
+
+	diskTracker := tbl.GetDiskTracker()
+	diskTracker.SetLabel(memory.LabelForCTEStorage)
+	diskTracker.AttachTo(ctx.GetSessionVars().StmtCtx.DiskTracker)
+
+	if config.GetGlobalConfig().OOMUseTmpStorage {
+		actionSpill := tbl.ActionSpill()
+		ctx.GetSessionVars().StmtCtx.MemTracker.FallbackOldAndSetNewAction(actionSpill)
+	}
+}
+
+func (e *CTEExec) tryFilterAndAdd(chk *chunk.Chunk,
+	storage cteutil.Storage,
+	hashtbl baseHashTable) (res *chunk.Chunk, err error) {
+	if e.isDistinct {
+		if chk, err = e.filterAndAddHashTable(chk, storage, hashtbl); err != nil {
+			return nil, err
+		}
+	}
+	if chk.NumRows() == 0 {
+		return chk, nil
+	}
+	return chk, storage.Add(chk)
+}
+
+func (e *CTEExec) filterAndAddHashTable(chk *chunk.Chunk,
```

Comment:
filterAndAddHashTable --> deduplicate

Reply:
done

Reply:
This function also add hash table. The new name is also ok. I will note this in comment.